staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic.go

```diff
@@ -0,0 +1,292 @@
+		backends:             syncedBackends{idMap: backends},
+		updater:              updater{},
+		lister:               informer.Lister(),
+		hasSynced:            informer.Informer().HasSynced,
+		webhookClientManager: cm,
+	}
+
+	// On any change, rebuild the config
+	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(_ interface{}) { manager.updateConfiguration() },
+		UpdateFunc: func(_, _ interface{}) { manager.updateConfiguration() },
+		DeleteFunc: func(_ interface{}) { manager.updateConfiguration() },
+	})
+
+	return manager, nil
+}
+
+type dynamic struct {
+	config               *Config
+	staticBackend        *Backend
+	backends             syncedBackends
+	updater              updater
+	lister               auditlisters.AuditSinkLister
+	hasSynced            func() bool
+	webhookClientManager webhook.ClientManager
+}
+
+type syncedBackends struct {
+	idMap map[types.UID]*Backend
+	sync.RWMutex
+}
+
+// updater forces configuration updates to run serially. This is a separate lock
+// so that events are not blocked while the backends build
+type updater struct{ sync.Mutex }
+
+// ProcessEnforcedEvents proccesses the given events per current backend map
+func (d *dynamic) ProcessEnforcedEvents(events ...*audit.EnforcedEvent) {
+	for _, b := range d.GetBackends() {
+		b.enforcedBackend.ProcessEnforcedEvents(events...)
+	}
+}
+
+// Run the dynamic backends
+func (d *dynamic) Run(stopCh <-chan struct{}) error {
+	// Shutdown channels are handled individually with the dynamic
+	// backend, message must be propagated out
+	go func() {
+		for {
+			select {
+			case <-stopCh:
+				d.Shutdown()
+				return
+			default:
+			}
+		}
+	}()
+	var funcs []func() error
+	for _, b := range d.GetBackends() {
+		funcs = append(funcs, func() error {
+			return b.enforcedBackend.Run(b.stopChan)
+		})
+	}
+	return errors.AggregateGoroutines(funcs...)
+}
+
+// Shutdown closes all backends stop channels and then runs their
+// subsequent shutdown methods
+func (d *dynamic) Shutdown() {
+	for _, b := range d.GetBackends() {
+		if !isChanClosed(b.stopChan) {
+			close(b.stopChan)
+		}
+		b.enforcedBackend.Shutdown()
+	}
+}
+
+// HasSynced tells whether the informer has synced
+func (d *dynamic) HasSynced() bool {
+	return d.hasSynced()
+}
+
+// GetBackends retrieves current backends in a safe manner
+func (d *dynamic) GetBackends() map[types.UID]*Backend {
+	d.backends.RLock()
+	defer d.backends.RUnlock()
+	return d.backends.idMap
+}
+
+// SetBackends sets the current backends in a safe manner
+func (d *dynamic) SetBackends(backends map[types.UID]*Backend) {
+	d.backends.Lock()
+	defer d.backends.Unlock()
+	d.backends.idMap = backends
+	return
+}
+
+// updateConfigurations is called from the shared informer and will compare the
+// current configurations found to the existing ones held. It will create/start
+// any new or updated backends, and remove/stop any deleted ones.
+func (d *dynamic) updateConfiguration() {
```

Comment:
I'm a little skeptical of this. Can you share a pointer to the code you copied this pattern from?

Reply:
I looked over the admission webhook implementation you referenced, and I think there is a key difference: In the admission case the webhooks are stateless, and `updateConfiguration` is actually just maintaining a list of the current configuration. That configuration is grabbed everytime a request needs to be sent off to the webhooks. The result is that `updateConfiguration` is much simpler in that case, as it's fine to just drop old configurations.

The problem here is that the backends are stateful, and need to be started & shutdown when added/removed. Because of that, the add & delete methods look quite different. So, I stand by my original suggestion: I think it would be cleaner to handle these 3 cases separately.

Reply:
Maybe this is my misunderstanding but can't the informer get out of sync? For an important function like auditing or admission wouldn't a consistency be the top priority?

Reply:
IIUC, that's what resync is for. See https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go#L48-L51

Reply:
ok @ncdc has informed me that infromers should never get out of sync for internal resource or that is a critical bug, I'll rework this